File: students/cbt/extract-questions.ts
import ZAI from 'z-ai-web-dev-sdk';
import fs from 'fs';
import path from 'path';

const SUBJECTS = ['physics', 'maths', 'english', 'economics'] as const;
type Subject = typeof SUBJECTS[number];

interface Question {
  id: number;
  year: string;
  question: string;
  options: string[];
  correctAnswer: number;
  hasImage?: boolean;
  imageFile?: string;
  explanation?: string;
}

// Initialize ZAI
let zai: Awaited<ReturnType<typeof ZAI.create>>;

async function initZAI() {
  zai = await ZAI.create();
  console.log('ZAI initialized');
}

async function analyzePage(imagePath: string, subject: string): Promise<string> {
  const imageBuffer = fs.readFileSync(imagePath);
  const base64Image = imageBuffer.toString('base64');
  
  const prompt = `You are an expert at extracting exam questions from educational materials.

Analyze this ${subject} exam page and extract ALL questions from it.

IMPORTANT RULES:
1. Extract EVERY question you can see on this page
2. Preserve the EXACT wording of questions as they appear in the image
3. For questions with diagrams/images, note "HAS_IMAGE: true" and describe what the diagram shows
4. Extract all options (A, B, C, D) exactly as written
5. Identify the correct answer if shown (usually indicated with an asterisk, bold, or answer key)
6. Note the year if visible on the page

OUTPUT FORMAT - Return a JSON array:
[
  {
    "year": "2023" or "Unknown",
    "question": "The exact question text as it appears",
    "options": ["Option A text", "Option B text", "Option C text", "Option D text"],
    "correctAnswer": 0-3 (index of correct option, or -1 if unknown),
    "hasImage": true/false,
    "imageDescription": "Description of diagram if hasImage is true"
  }
]

Return ONLY valid JSON array, no other text. If no questions found on this page, return [].`;

  const response = await zai.chat.completions.createVision({
    messages: [
      {
        role: 'user',
        content: [
          { type: 'text', text: prompt },
          { 
            type: 'image_url', 
            image_url: { 
              url: `data:image/png;base64,${base64Image}` 
            } 
          }
        ]
      }
    ],
    thinking: { type: 'disabled' }
  });

  return response.choices[0]?.message?.content || '[]';
}

function parseQuestions(jsonStr: string): Question[] {
  try {
    // Clean up the response - extract JSON if wrapped in markdown
    let cleanJson = jsonStr.trim();
    if (cleanJson.startsWith('```')) {
      cleanJson = cleanJson.replace(/```json?\n?/g, '').replace(/```/g, '').trim();
    }
    
    const parsed = JSON.parse(cleanJson);
    if (!Array.isArray(parsed)) return [];
    
    return parsed.map((q, index) => ({
      id: index,
      year: q.year || 'Unknown',
      question: q.question || '',
      options: Array.isArray(q.options) ? q.options : [],
      correctAnswer: typeof q.correctAnswer === 'number' ? q.correctAnswer : -1,
      hasImage: q.hasImage || false,
      imageDescription: q.imageDescription || undefined
    }));
  } catch (e) {
    console.error('Failed to parse JSON:', e);
    return [];
  }
}

async function processSubject(subject: Subject): Promise<Question[]> {
  const pagesDir = path.join(__dirname, 'pdf_pages', subject);
  const allQuestions: Question[] = [];
  let questionId = 1;
  
  if (!fs.existsSync(pagesDir)) {
    console.log(`No pages directory for ${subject}`);
    return [];
  }
  
  const pages = fs.readdirSync(pagesDir)
    .filter(f => f.endsWith('.png'))
    .sort((a, b) => {
      const numA = parseInt(a.match(/\d+/)?.[0] || '0');
      const numB = parseInt(b.match(/\d+/)?.[0] || '0');
      return numA - numB;
    });
  
  console.log(`\nProcessing ${subject}: ${pages.length} pages`);
  
  for (let i = 0; i < pages.length; i++) {
    const page = pages[i];
    const pagePath = path.join(pagesDir, page);
    
    console.log(`  Page ${i + 1}/${pages.length}: ${page}`);
    
    try {
      const result = await analyzePage(pagePath, subject);
      const questions = parseQuestions(result);
      
      for (const q of questions) {
        q.id = questionId++;
        allQuestions.push(q);
      }
      
      console.log(`    Found ${questions.length} questions`);
      
      // Small delay to avoid rate limiting
      await new Promise(r => setTimeout(r, 500));
    } catch (e) {
      console.error(`  Error processing ${page}:`, e);
    }
  }
  
  return allQuestions;
}

function generateJSFile(questions: Question[], subject: string): string {
  const subjectCapitalized = subject.charAt(0).toUpperCase() + subject.slice(1);
  
  let js = `// ${subjectCapitalized} Questions extracted from PDF
// Total questions: ${questions.length}
// Generated on: ${new Date().toISOString()}

export const ${subject}Questions = [
`;

  for (const q of questions) {
    js += `  {
    id: ${q.id},
    year: "${q.year}",
    question: "${q.question.replace(/"/g, '\\"').replace(/\n/g, '\\n')}",
    options: ${JSON.stringify(q.options.map(o => o.replace(/"/g, '\\"')))},
    correctAnswer: ${q.correctAnswer},${q.hasImage ? `
    hasImage: true,` : ''}
  },
`;
  }

  js += `];

export default ${subject}Questions;
`;
  
  return js;
}

async function main() {
  await initZAI();
  
  const results: Record<Subject, Question[]> = {
    physics: [],
    maths: [],
    english: [],
    economics: []
  };
  
  for (const subject of SUBJECTS) {
    console.log(`\n${'='.repeat(50)}`);
    console.log(`Processing ${subject.toUpperCase()}`);
    console.log('='.repeat(50));
    
    const questions = await processSubject(subject);
    results[subject] = questions;
    
    // Generate JS file
    const jsContent = generateJSFile(questions, subject);
    const outputPath = path.join(__dirname, 'js', `${subject}-questions.js`);
    fs.writeFileSync(outputPath, jsContent);
    
    console.log(`\nGenerated ${questions.length} questions for ${subject}`);
    console.log(`Saved to: ${outputPath}`);
    
    // Save raw JSON as backup
    const jsonPath = path.join(__dirname, 'pdfs', `${subject}-questions.json`);
    fs.writeFileSync(jsonPath, JSON.stringify(questions, null, 2));
    console.log(`JSON backup: ${jsonPath}`);
  }
  
  console.log('\n' + '='.repeat(50));
  console.log('EXTRACTION COMPLETE');
  console.log('='.repeat(50));
  
  for (const subject of SUBJECTS) {
    console.log(`${subject}: ${results[subject].length} questions`);
  }
}

main().catch(console.error);

File: students/cbt/extract_questions.py
#!/usr/bin/env python3
"""
CBT Question Extraction Script
Extracts questions from PDF files with proper formatting and image associations
"""

import pdfplumber
import json
import re
import os
from pathlib import Path

CBT_DIR = Path('/home/z/my-project/students/cbt')
PDF_DIR = CBT_DIR / 'pdfs'
JS_DIR = CBT_DIR / 'js'
IMAGES_DIR = CBT_DIR / 'images'

def extract_text_with_layout(pdf_path):
    """Extract text from PDF with layout preservation"""
    questions = []
    
    with pdfplumber.open(pdf_path) as pdf:
        current_year = "Unknown"
        
        for page_num, page in enumerate(pdf.pages, 1):
            text = page.extract_text()
            if not text:
                continue
            
            # Check for year markers
            year_match = re.search(r'(20\d{2})\s*(JAMB|WAEC|NECO|UTME)', text, re.IGNORECASE)
            if year_match:
                current_year = year_match.group(1)
            
            # Extract questions from page
            page_questions = parse_questions_from_text(text, current_year, page_num)
            questions.extend(page_questions)
    
    return questions

def parse_questions_from_text(text, year, page_num):
    """Parse questions from extracted text"""
    questions = []
    
    # Split by question numbers
    # Pattern 1: "1." or "1)" or "1:"
    # Pattern 2: Question followed by options
    
    lines = text.split('\n')
    current_question = None
    current_options = []
    current_text = []
    
    for line in lines:
        line = line.strip()
        if not line:
            continue
        
        # Check for question number
        q_match = re.match(r'^(\d+)[.)\]:]\s*(.+)', line)
        
        if q_match:
            # Save previous question if exists
            if current_question and current_options:
                questions.append({
                    'year': year,
                    'question': current_question,
                    'options': current_options,
                    'correctAnswer': -1,
                    'page': page_num
                })
            
            # Start new question
            current_question = q_match.group(2)
            current_options = []
            current_text = [q_match.group(2)]
        
        # Check for options
        elif current_question:
            opt_match = re.match(r'^([A-D])[.)\]:]\s*(.+)', line, re.IGNORECASE)
            if opt_match:
                current_options.append(opt_match.group(2))
            else:
                # Append to question text
                current_question += ' ' + line
    
    # Save last question
    if current_question and current_options:
        questions.append({
            'year': year,
            'question': current_question,
            'options': current_options,
            'correctAnswer': -1,
            'page': page_num
        })
    
    return questions

def process_subject(subject):
    """Process a single subject PDF"""
    pdf_path = PDF_DIR / f'{subject}.pdf'
    
    if not pdf_path.exists():
        print(f"PDF not found: {pdf_path}")
        return []
    
    print(f"\nProcessing {subject}...")
    
    questions = extract_text_with_layout(pdf_path)
    
    # Assign IDs
    for i, q in enumerate(questions, 1):
        q['id'] = i
    
    print(f"  Extracted {len(questions)} questions")
    
    return questions

def generate_js_file(questions, subject):
    """Generate JavaScript question file"""
    js_path = JS_DIR / f'{subject}-questions.js'
    
    js_content = f"""// {subject.capitalize()} Questions - Extracted from PDF
// Total: {len(questions)} questions
// Generated: {__import__('datetime').datetime.now().isoformat()}

export const {subject}Questions = [
"""
    
    for q in questions:
        options_str = ', '.join([f'"{escape_js(o)}"' for o in q['options']])
        js_content += f"""  {{
    id: {q['id']},
    year: "{escape_js(q['year'])}",
    question: "{escape_js(q['question'])}",
    options: [{options_str}],
    correctAnswer: {q['correctAnswer']}
  }},
"""
    
    js_content += """];

export default {subject}Questions;
""".replace('{subject}', subject)
    
    with open(js_path, 'w', encoding='utf-8') as f:
        f.write(js_content)
    
    print(f"  Generated: {js_path}")
    
    # Also save JSON backup
    json_path = PDF_DIR / f'{subject}-extracted.json'
    with open(json_path, 'w', encoding='utf-8') as f:
        json.dump(questions, f, indent=2, ensure_ascii=False)
    
    print(f"  JSON backup: {json_path}")

def escape_js(text):
    """Escape text for JavaScript string"""
    if not text:
        return ""
    return (text
        .replace('\\', '\\\\')
        .replace('"', '\\"')
        .replace('\n', '\\n')
        .replace('\r', '')
        .replace('\t', '\\t'))

def main():
    print("CBT Question Extraction")
    print("=" * 50)
    
    subjects = ['physics', 'maths', 'english', 'economics']
    
    total = 0
    for subject in subjects:
        questions = process_subject(subject)
        if questions:
            generate_js_file(questions, subject)
            total += len(questions)
    
    print("\n" + "=" * 50)
    print(f"Total questions extracted: {total}")

if __name__ == '__main__':
    main()
